File: foundations/UFO-Compression-Swift/tokenizer_coder.swift
// Watermark: ip zymatica.space
// Patent Pending — USPTO Provisional Application | Zymatica Project

import Foundation

public func writeVarint(_ val: Int) -> [UInt8] {
    var res = [UInt8]()
    var v = val
    while v >= 128 {
        res.append(UInt8((v & 0x7F) | 0x80))
        v >>= 7
    }
    res.append(UInt8(v & 0x7F))
    return res
}

public func readVarint(_ data: [UInt8], _ pos: inout Int) -> Int {
    var val = 0
    var shift = 0
    while true {
        if pos >= data.count {
            break
        }
        let b = data[pos]
        pos += 1
        val |= Int(b & 0x7F) << shift
        if (b & 0x80) == 0 {
            break
        }
        shift += 7
    }
    return val
}

/**
 * Level 4 Prefix-Suffix Vocabulary String Compression
 */
public func compressVocab(_ tokens: [[UInt8]]) -> [UInt8] {
    var encoded = [UInt8]()
    var prev = [UInt8]()
    for t in tokens {
        var common = 0
        let l = min(t.count, prev.count)
        while common < l && t[common] == prev[common] {
            common += 1
        }
        let suffix = Array(t[common...])
        encoded.append(contentsOf: writeVarint(common))
        encoded.append(contentsOf: writeVarint(suffix.count))
        encoded.append(contentsOf: suffix)
        prev = t
    }
    return encoded
}

/**
 * Level 4 Prefix-Suffix Vocabulary String Restoration
 */
public func decompressVocab(_ data: [UInt8], _ numTokens: Int) -> [[UInt8]] {
    var tokens = [[UInt8]]()
    var pos = 0
    var prev = [UInt8]()
    for _ in 0..<numTokens {
        if pos >= data.count {
            break
        }
        let common = readVarint(data, &pos)
        let suffixLen = readVarint(data, &pos)
        if pos + suffixLen > data.count {
            break
        }
        let suffix = Array(data[pos..<(pos + suffixLen)])
        pos += suffixLen
        
        var t = Array(prev[0..<min(common, prev.count)])
        t.append(contentsOf: suffix)
        tokens.append(t)
        prev = t
    }
    return tokens
}

/**
 * Level 3 BPE Merges Binary Index-Packing (24-bit integer pairs)
 */
public func compressMerges(_ merges: [(UInt32, UInt32)]) -> [UInt8] {
    var encoded = [UInt8]()
    encoded.reserveCapacity(merges.count * 6)
    for pair in merges {
        let idx0 = pair.0
        let idx1 = pair.1
        
        encoded.append(UInt8((idx0 >> 16) & 0xFF))
        encoded.append(UInt8((idx0 >> 8) & 0xFF))
        encoded.append(UInt8(idx0 & 0xFF))
        
        encoded.append(UInt8((idx1 >> 16) & 0xFF))
        encoded.append(UInt8((idx1 >> 8) & 0xFF))
        encoded.append(UInt8(idx1 & 0xFF))
    }
    return encoded
}

/**
 * Level 3 BPE Merges Binary Index-Unpacking (24-bit integer pairs)
 */
public func decompressMerges(_ data: [UInt8]) -> [(UInt32, UInt32)] {
    let numMerges = data.count / 6
    var merges = [(UInt32, UInt32)]()
    merges.reserveCapacity(numMerges)
    for i in 0..<numMerges {
        let offset = i * 6
        let idx0 = (UInt32(data[offset]) << 16) |
                   (UInt32(data[offset + 1]) << 8) |
                   UInt32(data[offset + 2])
                   
        let idx1 = (UInt32(data[offset + 3]) << 16) |
                   (UInt32(data[offset + 4]) << 8) |
                   UInt32(data[offset + 5])
        merges.append((idx0, idx1))
    }
    return merges
}

/**
 * Level 7 XOR-FEC Parity computation for error resilient transmission
 */
public func computeXorFecParity(_ chunks: [[UInt8]], _ chunkSize: Int) -> [UInt8] {
    var parity = [UInt8](repeating: 0, count: chunkSize)
    for chunk in chunks {
        let limit = min(chunk.count, chunkSize)
        for j in 0..<limit {
            parity[j] ^= chunk[j]
        }
    }
    return parity
}

File: foundations/UFO-Compression-Rust/tokenizer_coder.rs
// Watermark: ip zymatica.space
// Patent Pending — USPTO Provisional Application | Zymatica Project

pub fn write_varint(val: usize) -> Vec<u8> {
    let mut res = Vec::new();
    let mut v = val;
    while v >= 128 {
        res.push(((v & 0x7F) | 0x80) as u8);
        v >>= 7;
    }
    res.push((v & 0x7F) as u8);
    res
}

pub fn read_varint(data: &[u8], pos: &mut usize) -> usize {
    let mut val = 0;
    let mut shift = 0;
    loop {
        if *pos >= data.len() {
            break;
        }
        let b = data[*pos];
        *pos += 1;
        val |= ((b & 0x7F) as usize) << shift;
        if (b & 0x80) == 0 {
            break;
        }
        shift += 7;
    }
    val
}

/// Level 4 Prefix-Suffix Vocabulary String Compression
pub fn compress_vocab(tokens: &[&[u8]]) -> Vec<u8> {
    let mut encoded = Vec::new();
    let mut prev: &[u8] = b"";
    for &t in tokens {
        let mut common = 0;
        let l = std::cmp::min(t.len(), prev.len());
        while common < l && t[common] == prev[common] {
            common += 1;
        }
        let suffix = &t[common..];
        encoded.extend(write_varint(common));
        encoded.extend(write_varint(suffix.len()));
        encoded.extend(suffix);
        prev = t;
    }
    encoded
}

/// Level 4 Prefix-Suffix Vocabulary String Restoration
pub fn decompress_vocab(data: &[u8], num_tokens: usize) -> Vec<Vec<u8>> {
    let mut tokens = Vec::with_capacity(num_tokens);
    let mut pos = 0;
    let mut prev = Vec::new();
    for _ in 0..num_tokens {
        if pos >= data.len() {
            break;
        }
        let common = read_varint(data, &mut pos);
        let suffix_len = read_varint(data, &mut pos);
        if pos + suffix_len > data.len() {
            break;
        }
        let suffix = &data[pos .. pos + suffix_len];
        pos += suffix_len;
        
        let mut t = prev[0..std::cmp::min(common, prev.len())].to_vec();
        t.extend_from_slice(suffix);
        tokens.push(t.clone());
        prev = t;
    }
    tokens
}

/// Level 3 BPE Merges Binary Index-Packing
pub fn compress_merges(merges: &[(u32, u32)]) -> Vec<u8> {
    let mut encoded = Vec::with_capacity(merges.len() * 6);
    for &(idx0, idx1) in merges {
        // Encode each index as a 3-byte big-endian integer (24-bit packing)
        encoded.push(((idx0 >> 16) & 0xFF) as u8);
        encoded.push(((idx0 >> 8) & 0xFF) as u8);
        encoded.push((idx0 & 0xFF) as u8);
        
        encoded.push(((idx1 >> 16) & 0xFF) as u8);
        encoded.push(((idx1 >> 8) & 0xFF) as u8);
        encoded.push((idx1 & 0xFF) as u8);
    }
    encoded
}

/// Level 3 BPE Merges Binary Index-Unpacking
pub fn decompress_merges(data: &[u8]) -> Vec<(u32, u32)> {
    let num_merges = data.len() / 6;
    let mut merges = Vec::with_capacity(num_merges);
    for i in 0..num_merges {
        let offset = i * 6;
        let idx0 = ((data[offset] as u32) << 16) 
                 | ((data[offset + 1] as u32) << 8) 
                 | (data[offset + 2] as u32);
                 
        let idx1 = ((data[offset + 3] as u32) << 16) 
                 | ((data[offset + 4] as u32) << 8) 
                 | (data[offset + 5] as u32);
        merges.push((idx0, idx1));
    }
    merges
}

/// Level 7 XOR-FEC Parity computation for error resilient transmission
pub fn compute_xor_fec_parity(chunks: &[Vec<u8>], chunk_size: usize) -> Vec<u8> {
    let mut parity = vec![0u8; chunk_size];
    for chunk in chunks {
        for j in 0..std::cmp::min(chunk.len(), chunk_size) {
            parity[j] ^= chunk[j];
        }
    }
    parity
}
